src/packages/version-check/src/index.ts

```diff
@@ -0,0 +1,291 @@
+  protected _logger;
+  protected _currentVersion;
+
+  constructor(
+    currentVersion: string,
+    config?: VersionCheckConfig,
+    logger?: Function
+  ) {
+    // Creates a new config, or reads existing from disk
+    this.ConfigManager = new Conf({
+      configName: process.env.TEST ? "testConfig" : "config", // config is the package default
+      defaults: {
+        ...VersionCheck.DEFAULTS,
+        ...config
+      }
+    });
+
+    // pulls the config out of the manager, lays optional props over top
+    this._config = { ...this.ConfigManager.get("config"), ...config };
+
+    // If config was passed in, save changes
+    if (config) this.saveConfig();
+
+    if (this.isValidSemver(currentVersion)) {
+      this._currentVersion = this.cleanSemver(currentVersion);
+    } else {
+      // Semver is invalid, turn off version check
+      this._currentVersion = false;
+      this._config.enabled = false;
+    }
+
+    this._logger = logger || console;
+  }
+
+  init() {
+    this.getLatestVersion();
+    return this;
+  }
+
+  cleanSemver(semver) {
+    return semverClean(semver);
+  }
+
+  isValidSemver(semver) {
+    return semverValid(semver);
+  }
+
+  setEnabled(enabled) {
+    this.set("enabled", enabled);
+  }
+
+  setLatestVersion(latestVersion) {
+    if (this.isValidSemver(latestVersion)) {
+      this.set("latestVersion", latestVersion);
+    }
+  }
+
+  setLatestVersionLogged(latestVersionLogged) {
+    this.set("latestVersionLogged", latestVersionLogged);
+  }
+
+  setPackageName(packageName) {
+    this.set("packageName", packageName);
+  }
+
+  setTTL(ttl) {
+    this.set("ttl", ttl);
+  }
+
+  setUrl(url) {
+    this.set("url", url);
+  }
+
+  private set(key, value) {
+    this._config[key] = value;
+    this.saveConfig();
+  }
+
+  private saveConfig() {
+    this.ConfigManager.set("config", this._config);
+  }
+
+  configFileLocation() {
+    return this.ConfigManager.path;
+  }
+
+  // Intentionally verbose here if we get logging involved it could aid debugging
+  canNotifyUser() {
+    const currentVersion = this._currentVersion;
+    const latestVersion = this._config.latestVersion;
+    // No currentVersion version passed in
+    if (!currentVersion) {
+      return false;
+    } else if (this.alreadyLoggedThisVersion()) {
+      return false;
+    }
+    // returns falsy if function cannot detect semver difference
+    return !!this.detectSemverChange(currentVersion, latestVersion);
+  }
+
+  detectSemverChange(currentVersion, latestVersion) {
+    if (
+      !currentVersion ||
+      !latestVersion ||
+      !this.isValidSemver(currentVersion) ||
+      !this.isValidSemver(latestVersion) ||
+      semverGte(currentVersion, latestVersion)
+    )
+      return null;
+
+    return semverDiff(currentVersion, latestVersion);
+  }
+
+  async getLatestVersion() {
+    if (!this._config.enabled) return false;
+    try {
+      const latestVersion = this.cleanSemver(await this.fetchLatestVersion());
```

Comment:
I don't think we should have to `clean` here, since we control the server and it should always produce a valid semver.

Reply:
:+1: https://github.com/trufflesuite/ganache/pull/3285/commits/3456cff60a5c1bde931a99c0fd5e9289f1acda2f#diff-d3c52bfdae42da5ce88cd5db96d272177e44283f952b3545816b03f08a2c8b85R139